tests/pyfunc/test_spark.py

```diff
@@ -1572,3 +1576,44 @@ def test_spark_udf_env_manager_with_invalid_pythonpath(
     )
 
     np.testing.assert_allclose(result, expected_pred_result, rtol=1e-5)
+
+
+def test_build_model_env(spark, sklearn_model, model_path, tmp_path, monkeypatch):
+    import sklearn
+
+    from mlflow.pyfunc.dbconnect_artifact_cache import extract_archive_to_dir
+
+    monkeypatch.setenv("DATABRICKS_RUNTIME_VERSION", "15.4.1")
+    model, inference_data = sklearn_model
+
+    with mlflow.start_run():
+        model_info = mlflow.sklearn.log_model(
+            model,
+            "model",
+            pip_requirements=[
+                f"scikit-learn=={sklearn.__version__}",
+                # `build_model_env` doesn't support building env with dev version MLflow,
+                # so add MLflow as a required dependency here.
+                "mlflow",
+            ],
+        )
+
+    model_uri = model_info.model_uri
+    model_env_path = build_model_env(model_uri, tmp_path)
+    archive_name = Path(model_env_path).name[:-7]
+    env_name = "-".join(archive_name.split("-")[:2])
+    extract_dir = Path("/tmp") / archive_name
+    try:
+        extract_archive_to_dir(model_env_path, extract_dir)
+        # Check the extracted python environment installs the expected sklearn package version.
+        subprocess.check_call(
+            [
+                "bash",
+                "-c",
+                f"source /tmp/{archive_name}/virtualenv_envs/{env_name}/bin/activate && "
+                f"python -c "
+                f"\"import sklearn; assert sklearn.__version__ == '{sklearn.__version__}'\"",
+            ]
+        )
+    finally:
+        shutil.rmtree(f"/tmp/{archive_name}", ignore_errors=True)
```

Comment:
can we test using `spark_udf`?

Reply:
No. Currently it uses `spark.addArtifact` and it only supports Databricks runtime, although OSS spark also have `spark.addArtifact`, but it has different behavior with Databricks runtime, OSS spark doesn't use NFS to store uploaded artifacts, and there is no `DB_SESSION_UUID` environmental variable in OSS spark)

Reply:
For now I don't plan to support OSS spark, because OSS users might use various kind of environment, prebuilding model env and shares it across different environment might cause unexpected issues, and cause user complaint.
so I don't support it currently, until I can fully figure out these issues.

Reply:
For e2e test, I will add test in databricks/universe.

Reply:
got it